src/Dotnet.Script.Core/ScriptPublisher.cs

```diff
@@ -0,0 +1,113 @@
+
+        public ScriptPublisher(ScriptProjectProvider scriptProjectProvider, ScriptCompiler scriptCompiler,
+            ScriptConsole scriptConsole)
+        {
+            _scriptProjectProvider = scriptProjectProvider ?? throw new ArgumentNullException(nameof(scriptProjectProvider));
+            _scriptCompiler = scriptCompiler ?? throw new ArgumentNullException(nameof(scriptCompiler));
+            _scriptConsole = scriptConsole ?? throw new ArgumentNullException(nameof(scriptConsole));
+            _scriptEnvironment = ScriptEnvironment.Default;
+        }
+
+        public ScriptPublisher(LogFactory logFactory, ScriptCompiler scriptCompiler)
+            : this
+            (
+                new ScriptProjectProvider(logFactory),
+                scriptCompiler,
+                ScriptConsole.Default
+            )
+        {
+        }
+
+        public void CreateExecutable(ScriptContext context, LogFactory logFactory)
+        {
+            var tempProjectPath = _scriptProjectProvider.CreateProjectForScriptFile(context.FilePath);
+            var tempProjectDirecory = Path.GetDirectoryName(tempProjectPath);
+
+            var scriptAssemblyPath = CreateScriptAssembly(context, tempProjectDirecory);
+
+            var projectFile = new ProjectFile(File.ReadAllText(tempProjectPath));
+            // todo: grab version in a better way?
+            projectFile.AddPackageReference(new PackageReference("Microsoft.CodeAnalysis.Scripting", "2.8.2", PackageOrigin.ReferenceDirective));
+            projectFile.AddAssemblyReference(scriptAssemblyPath);
+            projectFile.Save(tempProjectPath);
+
+            CopyProgramTemplate(tempProjectDirecory);
+
+            var runtimeIdentifier = _scriptEnvironment.RuntimeIdentifier;
+
+            var commandRunner = new CommandRunner(logFactory);
+            // todo: may want to add ability to return dotnet.exe errors
+            var exitcode = commandRunner.Execute("dotnet", $"publish \"{tempProjectPath}\" -c Release -r {runtimeIdentifier} -o {context.WorkingDirectory}");
+            if (exitcode != 0) throw new Exception($"dotnet publish failed with result '{exitcode}'");
+        }
+
+        private string CreateScriptAssembly(ScriptContext context, string tempProjectDirecory)
+        {
+            try
+            {
+                var compilationContext = _scriptCompiler.CreateCompilationContext<int, CommandLineScriptGlobals>(context);
+                var scriptOptions = compilationContext.Script.GetCompilation().Options
+                    .WithScriptClassName(AssemblyName);
+                var scriptCompilation = compilationContext.Script.GetCompilation()
+                      .WithOptions(scriptOptions)
+                    .WithAssemblyName(AssemblyName);
+
+                foreach (var reference in scriptCompilation.DirectiveReferences)
+                {
+                    var refInfo = new FileInfo(reference.Display);
```

Comment:
you may want to check for `nuget` pragma here, if it's a nuget reference there is nothing to copy (it resolves to a fake DLL) so you can skip it.

Reply:
k I'll add something for this, I noticed what you're talking about